SPEC

```diff
@@ -113,6 +113,25 @@ be implemented by the server.
                         fatal(message, &block)
 <tt>rack.multipart.buffer_size</tt>:: An Integer hint to the multipart parser as to what chunk size to use for reads and writes.
 <tt>rack.multipart.tempfile_factory</tt>:: An object responding to #call with two arguments, the filename and content_type given for the multipart form field, and returning an IO-like object that responds to #<< and optionally #rewind. This factory will be used to instantiate the tempfile for each multipart form file upload field, rather than the default class of Tempfile.
+
+Servers that choose to support WebSockets and SSE connections should follow these additional environment specifications:
+<tt>rack.upgrade?</tt>:: Is nil or missing if no connection upgrade is possible,
+                         :websocket for a WebSocket upgrade, or :sse for an SSE upgrade.
+<tt>rack.upgrade</tt>:: Is used to pass a handler object back to the server for WebSocket and SSE connections.
+                     The handler MAY implement any or all of the following callbacks:
+                      on_open(client)              # called when the connection is complete with a client object.
+                      on_message(client, message)  # called when a WebSocket message is received by server. <tt>message</tt> will be UTF-8 encoded for text messages and Binary encoded for binary messages.
+                      on_shutdown(client)          # may be called before a connection is closed due to server shutdown.
+                      on_close(client)             # called after the connection is closed.
+                      on_drained(client)           # may be called when the number of pending writes drops to zero.
+                     The <tt>client</tt> is used for writing and checking status of the upgraded connect. It has these methods.
+                      write(message)               # writes to the WebSocket or SSE connection
+                      close()                      # forces a close of the WebSocket or SSE connection
```

Comment:
If you want to have that, implement it as `close_write` which is what Ruby calls `shutdown(SHUT_WR)`. `close` should guarantee after returning that the underlying socket is closed. Otherwise, you are in for a world of pain.

Reply:
Agoo treats a cal to close as a request just like a write and places the close on the queue.

Reply:
@ohler55, so does iodine. I place a "close" marker on the packet queue, so `close` is always performed after all scheduled data is sent.


@ioquatix , I think your missing the point o f abstracting away the network and the protocol.

My suggestion was about clarifying this little bit, not changing what both iodine and agoo already implement.

We aren't authoring a network layer. We are authoring an abstracted application side API.

The reasonable exception is that `write` is perfumed before `close`. i.e., if my code is:

```ruby
write "hello"
close
```

The reasonable expectation is that "hello" is actually written.

There's no `force_close` or `close_write` in the specification because the application shouldn't be concerned with these things. If the application doesn't want data written, it can avoid writing it.

Reply:
You can make close call `flush`, or you can flush after every write. But you make close call flush, you better be careful about `EPIPE`.

Reply:
@ioquatix 

I think these are implementation details that each server could work out according to it's architecture.

The falcon implementation, for example, seems to block `write` so it always completes before `close` (or so I understand) while iodine's implementation places a flag and adds an event to the event queue.

Reply:
Sure as long as there are no pratical differences to the user, it would be reasonable. However, if you are buffering events, you might end up with the user having the wrong expectations.

```
10.times do
    connection.send_message(...)
end

# got here, were all events sent? it depends on implementation. Can I call flush to confirm they were sent?
```

Reply:
I'm not sure what you expect `flush` to do. In most uses flush blocks until all outstanding writes have completed. Is that the intent, block until completed? The PR proposes the `on_drained` callback for that purpose or a `pending` count of zero. Are you proposing a `wait_for_all_writes_to_complete` method called `flush` in addition to the existing ones?

Reply:
For a high level protocol like this, calling flush after each write would make sense to me.

It provides the user with a strong expectation, that after calling write, the data has been sent, and pending a network failure, would arrive, or else the write fails right then, with, say, `EPIPE`. Otherwise you'll just end up with a spaghetti state machine trying to handle all these conditions.

Reply:
@ioquatix ,

I hope I don't seem too blunt or crude. I very much appreciate the interest and willingness to polish the specification and make it both as clear and as practical as can be.

However, please consider the model to be totally separate from the network - there is no network. There's only this API.

We can change the API if we need features, but we don't expose network bits or logic because part of our job is to abstract these things away - so there is no network, there is no protocol (as much as possible).

In this sense, `flush` doesn't exist. It's a network / server detail that the client never sees, abstracted away by the server.

The closest an application can come to ask about these things is to ask about all the `pending` events that haven't yet completed. This allows an application to know if the `on_drain` callback is somewhere in their future.

The `pending` query doesn't to expose the network, it exposes the progress of existing API calls. This provides important information about the possibility of a slow client (or a slow clearing "queue") allowing an application to stop serving a resource hungry "client".

Reply:
If I understand your response you are answering yes to the question of whether or not a `flush` method should be added. Then in any application you wrote you would block until the write completes instead of making use of the `on_drained` callback. That is your choice of course.

Reply:
> However, please consider the model to be totally separate from the network - there is no network. There's only this API.

Fair enough.

Reply:
> If I understand your response you are answering yes to the question of whether or not a flush method should be added. Then in any application you wrote you would block until the write completes instead of making use of the on_drained callback. That is your choice of course.

I really find the inverted flow control of callback style programming horrible. So, I prefer `#flush` over an `#on_drained` callback. Callbacks = state machine spaghetti = hard to maintain/buggy code. It's just my personal opinion, FYI.